pkg/doc/verifiable/common.go

```diff
@@ -55,16 +55,47 @@ func (ja JWSAlgorithm) jose() (jose.SignatureAlgorithm, error) {
 // Proof defines embedded proof of Verifiable Credential
 type Proof interface{}
 
-type typedID struct {
+// TypedIDExtraFields is a map of extra fields which extend typed ID field.
+type TypedIDExtraFields map[string]interface{}
+
+// TypedID defines a flexible structure with id and name fields and arbitrary extra fields
+// kept in TypedIDExtraFields.
+type TypedID struct {
 	ID   string `json:"id,omitempty"`
 	Type string `json:"type,omitempty"`
+
+	TypedIDExtraFields `json:"-"`
 }
 
-// RefreshService provides a way to automatic refresh of expired Verifiable Credential
-type RefreshService typedID
+// MarshalJSON defines custom marshalling of TypedID to JSON.
+func (tid *TypedID) MarshalJSON() ([]byte, error) {
+	type Alias TypedID
+	alias := (*Alias)(tid)
+
+	data, err := marshalWithExtraFields(alias, tid.TypedIDExtraFields)
+	if err != nil {
+		return nil, fmt.Errorf("marshal TypedID: %w", err)
+	}
+
+	return data, nil
+}
 
-// TermsOfUse represents terms of use of Verifiable Credential by Issuer or Verifiable Presentation by Holder.
-type TermsOfUse interface{}
+// UnmarshalJSON defines custom unmarshalling of TypedID from JSON.
+func (tid *TypedID) UnmarshalJSON(data []byte) error {
+	type Alias TypedID
+	alias := (*Alias)(tid)
+
+	ef := make(TypedIDExtraFields)
+	err := unmarshalWithExtraFields(data, alias, ef)
+	if err != nil {
+		return fmt.Errorf("unmarshal TypedID: %w", err)
+	}
+
+	*tid = *(*TypedID)(alias)
```

Comment:
what's this?

Reply:
it seems UnmarshalJSON changes the value of the wrapped TypedID struct.. this is an update function, it does not do `only` unmarshal

Reply:
Yes, that's right. We would like to handle unmarshalling of TypedID in a different to standard way as want to keep extra fields map. `json` package allows doing so by the implementation of `json.Unmarshaler`. We had to add aliases trick at the very beginning to avoid infinite recursion (as `unmarshalWithExtraFields()` calls `json.Unmarshall()` on typedID which in turn calls `TypedID::UnmarshalJSON` etc.

Will it be clearer if we put the following assignments:
```
tid.ID = alias.ID
tid.Type = alias.Type
```
instead of
```
*tid = *(*TypedID)(alias)
```

Reply:
in either case, this is an update operation on `tid`, the meaning of Unmarshal is to take a marshalled content and unmarshal it `without` modifying the struct.

Reply:
I think the opposite - we are doing unmarshalling and should update the struct. Otherwise, there will be no effect of the operation - we return just `error` and take no arguments (which we could update).

Reply:
hey, just understood how stupid was that line of code
removed it